aten/src/ATen/native/mobile/cpu/internal/Convolution.cpp

```diff
@@ -0,0 +1,349 @@
+         // Groups
+         (groups > 0) &&
+         // Transpose
+         !transposed &&
+         // Input
+         (weight.size(Layout::Filter::Input) > 0) &&
+         // Output
+         (weight.size(Layout::Filter::Output) > 0) &&
+         // Output - Groups
+         ((weight.size(Layout::Filter::Output) % groups) == 0) &&
+         // Output Min / Max
+         (output_min.isIntegral(true) || output_min.isFloatingPoint()) &&
+         (output_max.isIntegral(true) || output_max.isFloatingPoint()) &&
+         (output_max.to<float>() > output_min.to<float>());
+}
+
+Context create(
+    const Tensor& weight,
+    const c10::optional<Tensor>& bias,
+    const IntArrayRef padding_,
+    const IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const int64_t groups,
+    const bool transposed,
+    const Scalar output_min,
+    const Scalar output_max) {
+  const auto padding = expand_param_if_needed(padding_, "padding", 2);
+  const auto stride = expand_param_if_needed(stride_, "stride", 2);
+  const auto dilation = expand_param_if_needed(dilation_, "dilation", 2);
+
+  TORCH_INTERNAL_ASSERT(
+      available(
+          weight,
+          bias,
+          padding,
+          stride,
+          dilation,
+          groups,
+          transposed,
+          output_min,
+          output_max),
+      "mobile::cpu::convolution not available!");
+
+  xnn_operator_t convolution_op{};
+
+  const xnn_status create_status = xnn_create_convolution2d_nhwc_f32(
+      padding[Layout::Parameter::Height],                               // input_padding_top
+      padding[Layout::Parameter::Width],                                // input_padding_right
+      padding[Layout::Parameter::Height],                               // input_padding_bottom
+      padding[Layout::Parameter::Width],                                // input_padding_left
+      weight.size(Layout::Filter::Height),                              // kernel_height
+      weight.size(Layout::Filter::Width),                               // kernel_width
+      stride[Layout::Parameter::Height],                                // subsampling_height
+      stride[Layout::Parameter::Width],                                 // subsampling_width
+      dilation[Layout::Parameter::Height],                              // dilation_height
+      dilation[Layout::Parameter::Width],                               // dilation_width
+      groups,                                                           // groups
+      weight.size(Layout::Filter::Input),                               // group_input_channels
+      weight.size(Layout::Filter::Output) / groups,                     // group_output_channels
+      weight.size(Layout::Filter::Input) * groups,                      // input_pixel_stride
+      weight.size(Layout::Filter::Output),                              // output_pixel_stride
+      weight.contiguous(MemoryFormat::ChannelsLast).data_ptr<float>(),  // kernel
```

Comment:
wait, how do you know the memory is still valid for the channels last contiguous version of weight?  There's nothing holding on to the Tensor reference.

Reply:
According to the C++ standard the lifetime of a temporary object is bound to the full-expression it is contained in.  In above usage, the call returns a prvalue whose lifetime is bound to the parent function call.

[Reference](https://en.cppreference.com/w/cpp/language/lifetime):
> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created ...

Relevant stackoverflow question: https://stackoverflow.com/questions/584824/guaranteed-lifetime-of-temporary-in-c

Having said that I will pull this out of the call to be clear.

Reply:
Done.

Reply:
I think(?) I'm asking something different.  I agree with your explanation about the call to xnn_create_convolution2d_nhwc_f32 -- that's perfectly safe.  But because you do this separate create step (in it's own function), and the Context only holds the data_pointer/sizes and not the tensor, how do you know the Tensor is around when you actually run the op?  I would have expected weight_nhwc to be around for the entire create_and_run step, not just create.

Reply:
Right, I see your point now and that is a great observation.  The short answer is that XNNPACK allocates an internal buffer and copies the weights over.  This internal buffer is contained within a struct which xnn_create_convolution2d_nhwc_f32 returns a handle to.

The longer answer is that convolution::Context in this file, holds onto an "Operator", which is basically a std::unique_ptr of an xnn_operator with a custom deleter - the same handle mentioned above.  xnn_operator in turn though, is a struct that holds onto [packed_weights](https://github.com/google/XNNPACK/blob/f46f67559607339870b77d210cae2f2434ed8f64/src/xnnpack/operator.h#L208), which is [allocated](https://github.com/google/XNNPACK/blob/f46f67559607339870b77d210cae2f2434ed8f64/src/convolution-nhwc.c#L622) (or other places in the same file depending on the type of the convolution operator) and [copied into](https://github.com/google/XNNPACK/blob/f46f67559607339870b77d210cae2f2434ed8f64/src/convolution-nhwc.c#L631) xnn_create_convolution2d_nhwc_f32.  Effectively we end up holding onto this struct, which contains the copied weights, with a unique_ptr as part of convolution::Context.

Reply:
ah ok, so the op basically assumes that the weight is in contiguous nhwc format so it knows what to copy over.